### ISSUE TITLE
Phalcon\Http\Client\Provider\Curl does not return 200 OK on 100 Continue requests.

### ISSUE THREAD

**Issue by cottton** [phalcon/incubator#684]
If a customer header is given then the response returns a 100 Continue http status.

Test:
- **calling** a controller that expects a post request
- using **heavy payload** to force server -> client 100 Continue
- call once **without** any custom **headers**
- **write** response into **file**
- call once **with** any custom **header**
- **write** response into **file**

Result:
_1st is expected, 2nd unexpected 100 Continue_
```
status code:	200
status message:	OK
body:
HTTP/1.1 200 OK
Server: nginx/1.11.5
Date: Tue, 01 Nov 2016 20:26:38 GMT
Content-Type: application/json; charset=utf-8
Transfer-Encoding: chunked
Connection: keep-alive
X-Powered-By: PHP/7.0.11

{"status":"OK","data":"kk","message":"OK","code":0}

--------------------

status code:	100
status message:	Continue
body:
HTTP/1.1 100 Continue

HTTP/1.1 200 OK
Server: nginx/1.11.5
Date: Tue, 01 Nov 2016 20:26:38 GMT
Content-Type: application/json; charset=utf-8
Transfer-Encoding: chunked
Connection: keep-alive
X-Powered-By: PHP/7.0.11

{"status":"OK","data":"kk","message":"OK","code":0}

--------------------
```

 **Test script**
```
<?php

require 'vendor/autoload.php';

/**
 * Prepare:
 */
// curl options
$options = [
    CURLOPT_RETURNTRANSFER  => true,
    CURLOPT_AUTOREFERER     => true,
    CURLOPT_FOLLOWLOCATION  => true,
    CURLOPT_MAXREDIRS       => 5,
    CURLOPT_HEADER          => true,
    CURLOPT_PROTOCOLS       => CURLPROTO_HTTP | CURLPROTO_HTTPS,
    CURLOPT_REDIR_PROTOCOLS => CURLPROTO_HTTP | CURLPROTO_HTTPS,
    CURLOPT_USERAGENT       => '',
    CURLOPT_CONNECTTIMEOUT  => 30,
    CURLOPT_TIMEOUT         => 30,
];

// A controller that expects a post request.
$uri = 'http://foo.bar/baz';

// Heavy payload to force server -> client 100 Continue
$parameter = array_fill(0, 10000, 'foo');

function debug($response)
{
    $content = "status code:\t{$response->header->statusCode}\r\n";
    $content .= "status message:\t{$response->header->statusMessage}\r\n";
    $content .= "body:\r\n{$response->body}\r\n";
    $content .= "\r\n--------------------\r\n\r\n";
    file_put_contents('data/response_debug.txt', $content, FILE_APPEND);
}


/**
 * Test: response returns 200 OK as expected:
 */
$curl = new Phalcon\Http\Client\Provider\Curl();
$curl->setOptions($options);

// Note: no headers to set.
$headers = [];

$response = $curl->post($uri, $parameter, true, $headers, false);
debug($response);


/**
 * Test: response returns 100 Continue unexpected:
 */
$curl = new Phalcon\Http\Client\Provider\Curl();
$curl->setOptions($options);

// Note: ANY header causes the 100 Continue status.
$headers = [];
foreach (
    [
        //'Accept'         => 'application/json',
        //'Accept-Charset' => 'utf-8',
        'X-Foo' => 'bar',
    ] as $key => $val
) {
    $headers[] = $key . ': ' . $val;
}

$response = $curl->post($uri, $parameter, true, $headers, false);
debug($response);

```

The problem is that a response is **not** **success** (< 200).

**Comment by sergeyklay** [comment]
Could you please provide Incubator version?

**Comment by cottton** [comment]
3.0.x

I just found out that `$customHeader` mean custom header =)

see https://github.com/phalcon/incubator/blob/master/Library/Phalcon/Http/Client/Provider/Curl.php#L116

`$header[] = 'Expect:';`

If i provide a `['Expect' => '']` as parameter `$customHeader` (as it happen if no custom header is given) then the correct status will be returned.

_Reason why it works: server does not return staus 100 AND 200 (or others). I did not know that the server (can/should) returns more than one HTTP/ line._

Is it intended to be so?

----------------------

So i guess i should use `$curl->header->addMultiple(['X-Foo' => 'bar']);` to add multiple headers.
But there is something weird: 

> Warning: array_combine(): Both parameters should have an equal number of elements in /var/www/html/vendor/phalcon/incubator/Library/Phalcon/Http/Client/Header.php on line 61

see: https://github.com/phalcon/incubator/blob/master/Library/Phalcon/Http/Client/Header.php#L61
Shouldnt it be `$this->fields = array_merge($this->fields, $fields);` ?
_Combine fields and fileds wont work ima_

**At least** i now use `$curl->header->set($name, $val);` to add each header. **Works fine.**

A bit more docu would be good :x 
Just sayin =)

**Comment by sergeyklay** [comment]
Well, `array_combine` creates an array by using one array for keys and another for its values. So I think we can improve addMultiple a bit more. Could you please try to sort out and send PR?

**Comment by cottton** [comment]
Done. see https://github.com/phalcon/incubator/pull/686

**Comment by cottton** [comment]
Created pull request for `$customHeader` -problem: https://github.com/phalcon/incubator/pull/688

Its ok, if previously set headers gets ignored. But we should get that `Expect:` fix working on custom headers also.

**Comment by sergeyklay** [comment]
Fixed in the `3.0.x` branch.
